File: Arrays and Linked List/Solutions/بالگرد/balgard.cpp
//by Mitra Omrani
#include<bits/stdc++.h>
using namespace std;
using ll=long long;
class node{
    friend class linklist;
    node(string nme,node * nx,node * prv){
        name=nme;
        next=nx;
        prev=prv;
    }
    string name;
    node * next;
    node * prev;
};
class linklist{
    int size;
    node * head;
    node * tail;

public:
    linklist(){
        size=0;
        head=0;
        tail=0;
    }
    ~linklist(){
          while(head){
            node *tmp= head->next;
            delete head;
            head=tmp;
        }
        size=0;
        head=nullptr;
        tail=nullptr;
    }
    void insert(string name,int pos){
        if(size==0)
        {
            node * tmp=new node(name,nullptr,nullptr);
            head=tmp;
            tmp->next=nullptr;
            tail=head;
            size++;
            return;
        }
        if(pos==0){
             node * tmp=new node(name,head,nullptr);
             head->prev=tmp;
             head=tmp;
             size++;
             return;
        }
        else{
            ll i=0;
            node * tmp=head;
            while(i+1!=pos){
                i++;
                tmp=tmp->next;
            }
            node *x=new node(name,tmp->next,tmp);
            if(tmp->next!=nullptr) tmp->next->prev=x;
            tmp->next=x;
            tail=x;
            size++;
        }
    }
    void depart(string nme){
        if(head==nullptr) return;
        size--;
        node * tmp=head;
        if(tmp->name==nme){
            head=head->next;
           if(head!=nullptr) head->prev=nullptr;
            delete tmp;
            return;
        }
        while (tmp->next!=nullptr)
        {
           if(tmp->next->name==nme){
               node * tt=tmp->next;
              if(tmp->next->next!=nullptr) tmp->next->next->prev=tmp;
               tmp->next=tmp->next->next;
               delete tt;
               return;
           }
           tmp=tmp->next;
        }  
    }
    void relocate(string nme,int mv){
        node * tmp=head;
        ll i=0;
        while(tmp){
            if(tmp->name==nme)
                break;
            tmp=tmp->next;
            i++;
        }
        depart(nme);
        insert(nme,i+mv);
    }  
    void reverse(ll i,ll j){
        node * tmp=head;
        ll k=1;
        while(k!=i){
            k++;
            tmp=tmp->next;
        }
        node * tt=tmp;
        while(k!=j){
            k++;
            tmp=tmp->next;
        }
        while( tmp!=tt && tt->next!=tmp )
        {
            string data=tt->name;
            tt->name=tmp->name;
            tmp->name=data;
            tt=tt->next;
            tmp=tmp->prev;
        }
        if(tt->next==tmp){
           string data=tt->name;
            tt->name=tmp->name;
            tmp->name=data;
        }
    }
    void print(){
        if(head==nullptr) {cout<<"The airport is empty"<<endl; return;}
        node * tmp=head;
        while (tmp)
        {
           cout<<tmp->name<<' ';
           tmp=tmp->next;
        }
        cout<<"\n";
    }
};
int main(){
    ll n,s,x,i,j;
    linklist helikopt;
    cin>>n>>s;
    string name="",command="";
    for(ll i=0;i<n;i++){
        cin>>name;
        helikopt.insert(name,i);
    }
    while(s--){
        cin>>command;
        if(command=="insert"){
            cin>>name>>x;
            helikopt.insert(name,x);
        }
        else if(command=="relocate"){
            cin>>name>>x;
            helikopt.relocate(name,x);
        }
        else if(command=="depart"){
            cin>>name;
            helikopt.depart(name);
        }
        else if(command=="reverse"){
            cin>>i>>j;
            if(i>j) x=i,i=j,j=x;
            helikopt.reverse(i,j);
        }
    }
      helikopt.print();
    return 0;
}

File: Arrays and Linked List/Solutions/بالگرد/balgard.py
#by Lida safari
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self, list):
        self.head = None
        self.tail = None
        self.size = 0
        for i in range(len(list)):
            self.push(list[len(list)-1-i])

    def find_index(self, name):
        tmp = self.head
        i = 1
        while tmp.data != name :
            if i > self.size:######################
                return #######################
            i += 1
            tmp = tmp.next

        return i

    def find_name(self, index):
        tmp = self.head
        while index-1:
            tmp = tmp.next
            index -= 1
        return tmp.data

    def push(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            self.tail = new_node
            self.size += 1
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node
            self.size += 1

    def insert(self, name, location):
        if not self.head:
            new_node = Node(name)
            self.head = new_node
            self.tail = new_node
            self.size += 1
        elif location == 0:
            self.push(name)
            return
        elif location == self.size:
            new_node = Node(name)
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
            self.size += 1
        else:
            new_node = Node(name)
            tmp = self.head
            for i in range(location):
                tmp = tmp.next
            tmp.prev.next = new_node
            new_node.prev = tmp.prev
            tmp.prev = new_node
            new_node.next=tmp
            self.size += 1

    def relocate(self, name, footpace):
        if footpace == 0:
            return
        tmp = self.head
        i = self.find_index(name)
        d = footpace + i - 1

        self.depart(name)

        if d < 0 :
            location = self.size+1+d
        else:
            location = d

        self.insert(name, location)

    def depart(self, name):
        if not self.head:
            return
        if self.size == 1:
            self.head = None
            self.size -= 1
            return
        index = self.find_index(name)
        if index == self.size:
            self.tail = self.tail.prev
            self.tail.next = None
        elif index == 1:
            self.head = self.head.next
            self.head.prev = None
        else:
            tmp = self.head
            for i in range(index-1):
                tmp = tmp.next
            tmp.prev.next = tmp.next
            tmp.next.prev = tmp.prev
        self.size -= 1

    def reverse(self, i, j):
        if i > self.size or j > self.size or i == j or i*j < 0:
            return
        tmp_first = self.head
        tmp_last = self.tail
        for z in range(i-1):
            tmp_first = tmp_first.next
        for z in range(self.size-j):
            tmp_last = tmp_last.prev
        for z in range((abs(j-i)+1)//2):
            tmp_last.data, tmp_first.data = tmp_first.data, tmp_last.data
            tmp_first = tmp_first.next
            tmp_last = tmp_last.prev
    def print(self):
        if not self.head:
            print("The airport is empty")
            return
        tmp = self.head
        while tmp.next:
            print(tmp.data, end=" ")
            tmp = tmp.next
        print(tmp.data)

s = input().split()
numb_H = int(s[0])
numb_O = int(s[1])
if numb_H != 0:
    h_list = input().split()
else:
    h_list = []
H_l_list = LinkedList(h_list)
for i in range(numb_O):
    order = input().split()
    if order[0] == "insert":
        H_l_list.insert(order[1], int(order[2]))
    if order[0] == "relocate":
        H_l_list.relocate(order[1], int(order[2]))
    if order[0] == "depart":
        H_l_list.depart(order[1])
    if order[0] == "reverse":
        H_l_list.reverse(int(order[1]), int(order[2]))
H_l_list.print()
